sky/spot/core.py

```diff
@@ -0,0 +1,299 @@
+"""SDK functions for managed spot job."""
+import os
+import tempfile
+from typing import Any, Dict, List, Optional, Union
+import uuid
+
+import colorama
+
+import sky
+from sky import backends
+from sky import exceptions
+from sky import sky_logging
+from sky import status_lib
+from sky import task as task_lib
+from sky.backends import backend_utils
+from sky.clouds.service_catalog import common as service_catalog_common
+from sky.skylet import constants as skylet_constants
+from sky.spot import constants
+from sky.spot import spot_utils
+from sky.usage import usage_lib
+from sky.utils import common_utils
+from sky.utils import controller_utils
+from sky.utils import dag_utils
+from sky.utils import rich_utils
+from sky.utils import subprocess_utils
+from sky.utils import ux_utils
+
+
+@usage_lib.entrypoint
+def queue(refresh: bool, skip_finished: bool = False) -> List[Dict[str, Any]]:
+    # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
+    """Get statuses of managed spot jobs.
+
+    Please refer to the sky.cli.spot_queue for the documentation.
+
+    Returns:
+        [
+            {
+                'job_id': int,
+                'job_name': str,
+                'resources': str,
+                'submitted_at': (float) timestamp of submission,
+                'end_at': (float) timestamp of end,
+                'duration': (float) duration in seconds,
+                'recovery_count': (int) Number of retries,
+                'status': (sky.spot.SpotStatus) of the job,
+                'cluster_resources': (str) resources of the cluster,
+                'region': (str) region of the cluster,
+            }
+        ]
+    Raises:
+        sky.exceptions.ClusterNotUpError: the spot controller is not up or
+            does not exist.
+        RuntimeError: if failed to get the spot jobs with ssh.
+    """
+    stopped_message = ''
+    if not refresh:
+        stopped_message = ('No in-progress spot jobs.')
+    try:
+        handle = backend_utils.is_controller_accessible(
+            controller_type=controller_utils.Controllers.SPOT_CONTROLLER,
+            stopped_message=stopped_message)
+    except exceptions.ClusterNotUpError as e:
+        if not refresh:
+            raise
+        handle = None
+        controller_status = e.cluster_status
+
+    if refresh and handle is None:
+        sky_logging.print(f'{colorama.Fore.YELLOW}'
+                          'Restarting controller for latest status...'
+                          f'{colorama.Style.RESET_ALL}')
+
+        rich_utils.force_update_status('[cyan] Checking spot jobs - restarting '
+                                       'controller[/]')
+        handle = sky.start(spot_utils.SPOT_CONTROLLER_NAME)
```

Comment:
Humm, seems like here we need to use the `_start` function in `sky/core.py`, but not `sky.start` since the latter will always return None. Should we update the return value of `sky.start` to make it return handle as well?

https://github.com/skypilot-org/skypilot/blob/48b8ca978e936412db757800b58450c19ae49976/sky/core.py#L818